### ISSUE TITLE
`LowStar.Buffer` specification is unsound under malloc(3) failure

### ISSUE THREAD

**Issue by landonf** [FStarLang/FStar#2114]
Functions that allocate new buffers in `LowStar.Monotonic.Buffer` currently fail to represent allocation failure in their specification.

Example:

```fstar
module Malloc

module HS = FStar.HyperStack
module B = LowStar.Buffer

let example ()
= let b = B.malloc HS.root 0UL 128ul in
  // If allocation fails, this is an arbitrary dereference relative to the NULL page
  B.index b 0ul
```

Generated C:

```C
uint64_t example()
{
  uint64_t *b = KRML_HOST_CALLOC((uint32_t)128U, sizeof (uint64_t));
  return b[0U];
}
```

kremlib's `KRML_HOST_(M|C)ALLOC` definitions:

```C
#define KRML_HOST_MALLOC malloc
#define KRML_HOST_CALLOC calloc
```

An obvious quick-fix would be to unconditionally abort on the `NULL` case, but given that heap exhaustion is an expected condition in some environments (especially low-level targets), would it be better to strengthen the post-conditions to support reasoning about `malloc(3)` failure?

I don't have a great answer for stack allocations, given that behavior under stack exhaustion is undefined, and statically reasoning about runtime stack use in a portable manner would be very difficult. While stack overflows *should* hit the `PROT_NONE` guard page(s) -- at which point a page fault handler will generally expand the stack or terminate the process -- there remains the risk of large stack allocations 'hopping' over the guard page region, which can then be used to stomp on reachable memory outside the thread's stack.

Would it be valuable to support reasoning about memory availability via a preorder relation over the heap state? For example, given a target where the size of the runtime stack guard page is known, limiting the total size of allocations within a stack frame could be used to ensure that they cannot exceed the span of the guard page region.

**Comment by msprotz** [comment]
Correct, `malloc` failures are something we're explicitly not modeling.

We've talked about it in the past; you could conceivably have:

```
val mmalloc_faillible (#a:Type0) (#rrel:srel a)
  (r:HS.rid) (init:a) (len:U32.t)
  :HST.ST (b:lmbuffer a rrel rrel (U32.v len){frameOf b == r /\ freeable b})
          (requires (fun _       -> malloc_pre r len))
          (ensures  (fun h0 b h1 -> not (g_is_null b) ==> alloc_post_mem_common b h0 h1 (Seq.create (U32.v len) init)))
```

in LowStar.Monotonic.Buffer, then modify https://github.com/FStarLang/FStar/blob/master/src/extraction/FStar.Extraction.Kremlin.fs#L792 to add `mmalloc_faillible` to the list. This would at least allow you to audit your source-code to figure out where you're not taking heap allocation failures into account. You could also author a wrapper that uses `LowStar.Failure` for allocation failures, allowing you to regain the same signature as the original `mmalloc`, but with the guarantee that the program aborts cleanly rather than continue its execution.

For Stack allocation, maybe this would be a good use-case for layered effects, where an extra ghost argument to the effect would contain a stack of upper bounds for allocations; every call to push_frame would push a zero on top of the stack, every call to alloca would increase the value on top of the stack, and every call to pop_frame would generate a proof obligation that the total size of allocations within the stack frame does not exceed the limit.

@aseemr do you think that would be feasible? I'd actually be quite keen on trying to work out an example

**Comment by aseemr** [comment]
For heap allocations, see the allocation functions with the `_partial` suffix in `LowStar.Monotonic.Buffer.fsti`, e.g. `mgcmalloc_partial`. It is currently not extracted via Kremlin, but that should be easy to add (a minor change in `Kremlin.fs`).

For stack allocations, we can discuss what do you have in mind with layered effects. Another way to do this would be to store a current stack allocation size in the memory model (along with the tip), and have the proof obligation as part of stack allocation functions, while manipulating this size appropriately with `push` and `pop` (along with the `tip` as is done currently).

But won't that requires `sizeof` which we don't have?

**Comment by msprotz** [comment]
Ah thanks Aseem I didn't know that the fallible functions were already written. Then it means we'd just have to add them at the location I mentioned in FStar.Extraction.Kremlin.fst

For stack allocation size, I was thinking of taking advantage of a recent kremlin optimization that erases pointers to units, so we could do something simple and define a global `let stack_size = B.pointer (erased (list nat))`. I haven't given this more thought so I'd probably need your help to flesh out a prototype.

That's a good point about `sizeof` -- I was thinking we could either define a sizeof predicate for base types (integers, booleans, units, etc.), or use a meta-programmed tactic that would do a conservative approximation of the size of a given type (e.g. struct) by doing some very liberal padding. This would also be useful to get rid of those pesky KRML_CHECK_SIZE in regular allocation functions.

**Comment by nikswamy** [comment]
I don't think we can meaningfully bound the stack size by only tracking the size of explicit stack allocations. The size of the call stack is below the level of abstraction of a Low* program. At best, if we were to add totality checks, we could say that a program uses a finite amount of stack.

**Comment by msprotz** [comment]
What worries me is the pervasive usage of VLA, i.e. passing a non-constant argument to `alloca`, so my thinking was that we could keep track of these variable-size allocations and put an upper bound on those, as a sanity check. I agree that we can't do stack usage analysis but putting a constant upper bound (i.e. as you say, stating that the amount of stack is finite) would rule these out, and would be a good first step.
